mod-clans/src/main/java/gg/solarmc/loader/clans/ClanManager.java

```diff
@@ -0,0 +1,177 @@
+
+
+public class ClanManager implements DataManager {
+
+    private final Cache<Integer,Clan> clans = Caffeine.newBuilder().expireAfterAccess(Duration.ofMinutes(10)).build();
+    private final Cache<Integer,Integer> allianceCache = Caffeine.newBuilder().build();
+
+    /**
+     * Gets a clan from memory cache or from table if not present.
+     * @param transaction the tx
+     * @param id ID of clan requested
+     * @return the clan you asked for idiot
+     * @throws IllegalStateException if the clan isn't present in table
+     */
+    public Clan getClan(Transaction transaction, Integer id) {
+        return clans.get(id, i -> {
+
+            var jooq = transaction.getProperty(DSLContext.class);
+
+            ClansClanInfoRecord rec = jooq.fetchOne(CLANS_CLAN_INFO,CLANS_CLAN_INFO.CLAN_ID.eq(i));
+
+            if (rec == null) {
+                throw new IllegalStateException("No such clan exists!");
+            }
+
+            Set<ClanMember> members = jooq.select(CLANS_CLAN_MEMBERSHIP.USER_ID).from(CLANS_CLAN_MEMBERSHIP)
+                    .where(CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(id)).fetchSet((rec1) -> new ClanMember(id,rec1.value1(),this));
+
+            ClansClanAlliancesRecord recAlly = jooq.fetchOne(CLANS_CLAN_ALLIANCES,CLANS_CLAN_ALLIANCES.CLAN_ID.eq(id));
+
+            ClanMember owner = new ClanMember(id,rec.getClanLeader(),this);
+
+            Clan returned = new Clan(rec.getClanId(), rec.getClanName(),rec.getClanKills(),
+                    rec.getClanDeaths(),rec.getClanAssists(),this,members,owner);
+
+            if (recAlly != null) {
+                this.insertAllianceCache(recAlly.getClanId(),recAlly.getAllyId());
+            }
+
+            return returned;
+
+        });
+    }
+
+    Optional<Integer> getAllyFromCache(Integer clanId) {
+        return Optional.ofNullable(allianceCache.getIfPresent(clanId));
+    }
+
+    /**
+     * Returns an optional representing the allied clan
+     * @param clan The clan you want to check for alliance
+     * @return the optional containing allied clan.
+     */
+
+    /*public Optional<Clan> s(Clan clan) {
+        return Optional.ofNullable(alliance.getIfPresent(clan));
+    }*/
+
+    /**
+     * Invalidates an alliance, not order sensitive
+     * Information for api users - this invalidates two rows in the cache.
+     * @param clan1 the first clan id, ally of the second
+     * @param clan2 the second clan id, ally of the first
+     */
+    void invalidateAllianceCache(Integer clan1, Integer clan2) {
+        this.allianceCache.invalidate(clan1);
+        this.allianceCache.invalidate(clan2);
+    }
+
+    /**
+     * (a248 please review this method, i'm not sure if inserting two rows here is an issue
+     * because it will be called on clan generation for the first and for the second clan,
+     * and i'm not sure if this will lead to 2 rows {cache ignores duplicate} or to 4 rows { cache accepts duplicate} )
```

Comment:
The cache acts like a map. If you insert a new value for the key, any previous value is replaced.

This method implementation is perfectly fine